src/compiler/program.ts

```diff
@@ -1454,7 +1569,89 @@ namespace ts {
+                if (kind === ExtensionKind.SemanticLint) {
+                    initializedLints[i] = {name: lints[i].name, walker: new (lints[i].ctor as SemanticLintProviderStatic)(ts, getTypeChecker(), lints[i].args), accepted: true};
+                }
+                else if (kind === ExtensionKind.SyntacticLint) {
+                    initializedLints[i] = {name: lints[i].name, walker: new (lints[i].ctor as SyntacticLintProviderStatic)(ts, lints[i].args), accepted: true};
+                }
+            }
+
+            visitNode(sourceFile);
+
+            return diagnostics;
+
+            function visitNode(node: Node) {
+                let oneAccepted = false;
+                const oldParent = parent;
+                const needsReset: Map<boolean> = {};
+                for (let i = 0; i < initializedLints.length; i++) {
+                    if (initializedLints[i].accepted) {
+                        activeLint = initializedLints[i];
+                        node.parent = parent;
+                        activeLint.walker.visit(node, stop, error);
+                        if (activeLint.accepted) {
+                            oneAccepted = true;
+                        }
+                        else {
+                            needsReset[i] = true;
+                        }
+                    }
+                }
+                parent = node;
+                if (oneAccepted) {
+                    forEachChild(node, visitNode);
+                }
+                parent = oldParent;
+                for (let i = 0; i < initializedLints.length; i++) {
+                    if (needsReset[i]) {
+                        initializedLints[i].accepted = true;
+                        needsReset[i] = false;
+                    }
+                }
+            }
+
+            function stop() {
+                activeLint.accepted = false;
+            }
+
+            function error(err: string): void;
+            function error(err: string, node: Node): void;
+            function error(err: string, start: number, length: number): void;
+            function error(err: string, nodeOrStart?: Node | number, length?: number): void {
+                if (typeof nodeOrStart === "undefined") {
+                    diagnostics.push(createExtensionDiagnostic(activeLint.name, err, sourceFile));
+                }
+                else if (typeof nodeOrStart === "number") {
+                    diagnostics.push(createExtensionDiagnostic(activeLint.name, err, sourceFile, nodeOrStart, length));
+                }
+                else {
+                    diagnostics.push(createExtensionDiagnosticForNode(nodeOrStart, activeLint.name, err));
+                }
+            }
+        }
+
         function getSyntacticDiagnosticsForFile(sourceFile: SourceFile, cancellationToken: CancellationToken): Diagnostic[] {
+            if (!(sourceFile.isDeclarationFile || sourceFile.externalModuleIndicator)) {

```

Comment:
Nope, though I would like us to add a mechanism like that in the future (e.g. parse and check files outside of rootDir, but don't emit them).